tesseract_command_language/include/tesseract_command_language/utils/utils.h

```diff
@@ -150,6 +151,97 @@ bool clampToJointLimits(Waypoint& wp,
  */
 CompositeInstruction generateSkeletonSeed(const CompositeInstruction& composite_instructions);
 
+/**
+ * @brief Convert a CompositeInstruction to delimited formate file by extracting all MoveInstructions
+ * @param composite_instructions The CompositeInstruction to extract data from
+ * @param file_path The location to save the file
+ * @param separator The separator to use
+ * @return true if successful
+ */
+bool toDelimitedFile(const CompositeInstruction& composite_instructions,
+                     const std::string& file_path,
+                     char separator = ',');
+
+/**
+ * @brief Equal operator for Waypoint Types
+ * @param lhs Waypoint
+ * @param rhs Waypoint
+ * @return True if equal, otherwise false
+ */
+bool operator==(const Waypoint& lhs, const Waypoint& rhs);
+
+/**
+ * @brief Not equal operator for Waypoint Types
+ * @param lhs Waypoint
+ * @param rhs Waypoint
+ * @return True if not equal, otherwise false
+ */
+bool operator!=(const Waypoint& lhs, const Waypoint& rhs);
+
+/**
+ * @brief Equal operator for MoveInstruction
+ * @param lhs MoveInstruction
+ * @param rhs MoveInstruction
+ * @return True if equal, otherwise false
+ */
+bool operator==(const MoveInstruction& lhs, const MoveInstruction& rhs);
+
+/**
+ * @brief Not equal operator for MoveInstruction
+ * @param lhs MoveInstruction
+ * @param rhs MoveInstruction
+ * @return True if not equal, otherwise false
+ */
+bool operator!=(const MoveInstruction& lhs, const MoveInstruction& rhs);
+
+/**
+ * @brief Equal operator for PlanInstruction
+ * @param lhs PlanInstruction
+ * @param rhs PlanInstruction
+ * @return True if equal, otherwise false
+ */
+bool operator==(const PlanInstruction& lhs, const PlanInstruction& rhs);
+
+/**
+ * @brief Not equal operator for PlanInstruction
+ * @param lhs PlanInstruction
+ * @param rhs PlanInstruction
+ * @return True if not equal, otherwise false
+ */
+bool operator!=(const PlanInstruction& lhs, const PlanInstruction& rhs);
+
+/**
+ * @brief Equal operator for CompositeInstruction
+ * @param lhs CompositeInstruction
+ * @param rhs CompositeInstruction
+ * @return True if equal, otherwise false
+ */
+bool operator==(const CompositeInstruction& lhs, const CompositeInstruction& rhs);
+
+/**
+ * @brief Not equal operator for CompositeInstruction
+ * @param lhs CompositeInstruction
+ * @param rhs CompositeInstruction
+ * @return True if not equal, otherwise false
+ */
+bool operator!=(const CompositeInstruction& lhs, const CompositeInstruction& rhs);
+
+/**
+ * @brief Equal operator for Instruction Types
+ * @param lhs Instruction
+ * @param rhs Instruction
+ * @return True if equal, otherwise false
+ */
+bool operator==(const Instruction& lhs, const Instruction& rhs);
+
+/**
+ * @brief Not equal operator for Instruction Types
+ * @param lhs Instruction
+ * @param rhs Instruction
+ * @return True if not equal, otherwise false
+ */
+bool operator!=(const Instruction& lhs, const Instruction& rhs);
```

Comment:
Why are these in utils.h but WaitInstruction for example has them in the class?

Reply:
Ah, in the case of comparing Waypoints it needs to cast back to the original type to do the comparison which pushed everything the utils to avoid Waypoint being aware of the other types. I have an idea and may be able to do this in the waypoint to avoid having them here.

Reply:
Everything except Waypoint and Instruction compare operator is located in the associated classes.